presto-hudi/src/main/java/com/facebook/presto/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -0,0 +1,106 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A runnable to load Hudi splits asynchronously in the background
+ */
+public class HudiBackgroundSplitLoader
+        implements Runnable
+{
+    private static final Logger log = Logger.get(HudiBackgroundSplitLoader.class);
+
+    private final ConnectorSession session;
+    private final ExtendedHiveMetastore metastore;
+    private final HudiTableLayoutHandle layout;
+    private final HoodieTableFileSystemView fsView;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final List<String> partitions;
+    private final String latestInstant;
+    private final int splitGeneratorNumThreads;
+    private final ExecutorService splitGeneratorExecutorService;
+
+    public HudiBackgroundSplitLoader(
+            ConnectorSession session,
+            ExtendedHiveMetastore metastore,
+            ExecutorService splitGeneratorExecutorService,
+            HudiTableLayoutHandle layout,
+            HoodieTableFileSystemView fsView,
+            AsyncQueue<ConnectorSplit> asyncQueue,
+            List<String> partitions,
+            String latestInstant)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.layout = requireNonNull(layout, "layout is null");
+        this.fsView = requireNonNull(fsView, "fsView is null");
+        this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
+        this.partitions = requireNonNull(partitions, "partitions is null");
+        this.latestInstant = requireNonNull(latestInstant, "latestInstant is null");
+
+        this.splitGeneratorNumThreads = getSplitGeneratorParallelism(session);
+        this.splitGeneratorExecutorService = requireNonNull(splitGeneratorExecutorService, "splitGeneratorExecutorService is null");
+    }
+
+    @Override
+    public void run()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+        List<HudiPartitionSplitGenerator> splitGeneratorList = new ArrayList<>();
+        List<Future> splitGeneratorFutures = new ArrayList<>();
+        ConcurrentLinkedQueue<String> concurrentPartitionQueue = new ConcurrentLinkedQueue<>(partitions);
+
+        // Start a number of partition split generators to generate the splits in parallel
+        for (int i = 0; i < splitGeneratorNumThreads; i++) {
+            HudiPartitionSplitGenerator generator = new HudiPartitionSplitGenerator(
+                    session, metastore, layout, fsView, asyncQueue, concurrentPartitionQueue, latestInstant);
+            splitGeneratorList.add(generator);
+            splitGeneratorFutures.add(splitGeneratorExecutorService.submit(generator));
+        }
+
+        // Wait for all split generators to finish
+        for (Future future : splitGeneratorFutures) {
+            try {
+                future.get();
+            }
+            catch (InterruptedException | ExecutionException e) {
```

Comment:
any other exception is also going to fail the split loader runnable right? i.e gets thrown back to HudiSplitSource?

Reply:
that's correct.